File: benchmarks/wrappers/gcp/python/nosql.py
from typing import List, Optional, Tuple

from google.cloud import datastore


class nosql:

    instance: Optional["nosql"] = None

    """
    Each benchmark supports up to two keys - one for grouping items,
    and for unique identification of each item.

    In Google Cloud Datastore, we determine different tables by using
    its value for `kind` name.

    The primary key is assigned to the `kind` value.

    To implement sorting semantics, we use the ancestor relation:
    the sorting key is used as the parent.
    It is the assumption that all related items will have the same parent.
    """

    def __init__(self, database: str):
        self._client = datastore.Client(database=database)

    def insert(
        self,
        table_name: str,
        primary_key: Tuple[str, str],
        secondary_key: Tuple[str, str],
        data: dict,
    ):

        parent_key = self._client.key(primary_key[0], primary_key[1])
        key = self._client.key(
            # kind determines the table
            table_name,
            # main ID key
            secondary_key[1],
            # organization key
            parent=parent_key,
        )

        val = datastore.Entity(key=key)
        val.update(data)
        self._client.put(val)

    def update(
        self,
        table_name: str,
        primary_key: Tuple[str, str],
        secondary_key: Tuple[str, str],
        data: dict,
    ):
        # There is no direct update - we have to fetch the entire entity and manually change fields.
        self.insert(table_name, primary_key, secondary_key, data)

    def get(
        self, table_name: str, primary_key: Tuple[str, str], secondary_key: Tuple[str, str]
    ) -> Optional[dict]:

        parent_key = self._client.key(primary_key[0], primary_key[1])
        key = self._client.key(
            # kind determines the table
            table_name,
            # main ID key
            secondary_key[1],
            # organization key
            parent=parent_key,
        )

        res = self._client.get(key)
        if res is None:
            return None

        # Emulate the kind key
        res[secondary_key[0]] = secondary_key[1]

        return res

    """
        This query must involve partition key - it does not scan across partitions.
    """

    def query(
        self, table_name: str, primary_key: Tuple[str, str], secondary_key_name: str
    ) -> List[dict]:

        ancestor = self._client.key(primary_key[0], primary_key[1])
        query = self._client.query(kind=table_name, ancestor=ancestor)
        res = list(query.fetch())

        # Emulate the kind key
        for item in res:
            item[secondary_key_name] = item.key.name

        return res

    def delete(self, table_name: str, primary_key: Tuple[str, str], secondary_key: Tuple[str, str]):
        parent_key = self._client.key(primary_key[0], primary_key[1])
        key = self._client.key(
            # kind determines the table
            table_name,
            # main ID key
            secondary_key[1],
            # organization key
            parent=parent_key,
        )

        return self._client.delete(key)

    @staticmethod
    def get_instance(database: Optional[str] = None):
        if nosql.instance is None:
            assert database is not None
            nosql.instance = nosql(database)
        return nosql.instance

File: benchmarks/wrappers/gcp/nodejs/nosql.js
// Copyright 2020-2025 ETH Zurich and the SeBS authors. All rights reserved.

const { Datastore, KEY } = require("@google-cloud/datastore");

/*
This implementation is the Node.js reimplementation of the reference
implementation in Python. It is used for the NoSQL benchmarks.

Each benchmark supports up to two keys - one for grouping items,
and for unique identification of each item.

In Google Cloud Datastore, we determine different tables by using
its value for `kind` name.

The primary key is assigned to the `kind` value.

To implement sorting semantics, we use the ancestor relation:
the sorting key is used as the parent.
It is the assumption that all related items will have the same parent.
*/

class nosql {
  constructor(database) {
    this._client = new Datastore({ 'databaseId': database });
  }

  _get_entity_key(table_name, primary_key, secondary_key) {
    return this._client.key([
      primary_key[0], primary_key[1], table_name, secondary_key[1]
    ]);
  }

  async insert(table_name, primary_key, secondary_key, data) {
    const key = this._get_entity_key(table_name, primary_key, secondary_key);
    await this._client.save({ key, data });
  }

  async update(table_name, primary_key, secondary_key, updates) {
    // Just like in the Python version, we don't have a direct update.
    // Instead, we fetch the existing data, update fields, and write it.
    // Otherwise, we would also rewrite fields that we do not want to modify.
    const key = this._get_entity_key(table_name, primary_key, secondary_key);
    let [res] = await this._client.get(key);
    if (!res) {
      res = {};
    }
    Object.assign(res, updates);
    await this._client.save({ key, data: res });
  }

  async get(table_name, primary_key, secondary_key) {
    const key = this._get_entity_key(table_name, primary_key, secondary_key);
    const [res] = await this._client.get(key);
    if (!res) {
      return null;
    }

    // Emulate the kind and main keys
    res[secondary_key[0]] = secondary_key[1];
    res[primary_key[0]] = primary_key[1];
    return res;
  }

  async query(table_name, primary_key, secondary_key_name) {
    const ancestor = this._client.key([primary_key[0], primary_key[1]]);
    const query = this._client.createQuery(table_name).hasAncestor(ancestor);
    const [res] = await this._client.runQuery(query);

    // Emulate the kind key
    for (const item of res) {
      item[secondary_key_name] = item[Datastore.KEY].name;
    }

    return res;
  }

  async delete(table_name, primary_key, secondary_key) {
    const key = this._get_entity_key(table_name, primary_key, secondary_key);
    await this._client.delete(key);
  }

  static get_instance(database = null) {
    // There's one database we connect to, so we can preallocate storage instance.
    if (!nosql.instance) {
      if (!database) {
        throw new Error("NoSQL database must be provided when creating an instance.");
      }
      nosql.instance = new nosql(database);
    }
    return nosql.instance;
  }
}
exports.nosql = nosql;
